File: 0801-is-graph-bipartite/0801-is-graph-bipartite.cpp
class Solution {
public:
vector<vector<int>> adjList;

bool dfs(int node, vector<int> &seen, int color) {
    seen[node] = color;
    
    for (auto nei: adjList[node]) {
        if (seen[nei] == 0) {
            if (!dfs(nei, seen, 3 - color)) return false;
        }
        else if(seen[nei] == color) return false;
    }
    
    return true;
}
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
    vector<int> seen(n, 0);
    
    adjList = graph;
    
    for (int i = 0; i < n; i++) {
        if (seen[i] == 0) {
            if (!dfs(i, seen, 1)) return false;
        }
    }
    
    return true;
    }
};

File: 0801-is-graph-bipartite/0801-is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:

        colors = [-1 for i in range(len(graph))] #-1 no color, 0 red, 1 blue
        

        def dfs(node, color):
            colors[node] = color

            for nei in graph[node]:
                if colors[nei] == -1:
                    if not dfs(nei, not color):
                        return False
                elif colors[nei] == color:
                    return False
            
            return True
        

        for i in range(len(graph)):
            if colors[i] == -1:
                if not dfs(i, 0):
                    return False
        
        return True




        